File: implementation/task_4/sources/src/Main.java
import graph.Edge;
import graph.Graph;
import graph.GraphFileParser;
import path.DjikstraPathFinder;

import java.io.IOException;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        Integer mode = null;
        String path_to_graph_data = null;
        Integer source = null;
        Integer destination = null;

        if (args.length != 4) {
            System.err.println("`mode` [1 for sequential, 2 for concurrent] " +
                    "`path/to/graph/data.txt` `source` `destination`");
            System.exit(1);
        } else {
            mode = Integer.parseInt(args[0]);
            path_to_graph_data = args[1];
            source = Integer.parseInt(args[2]);
            destination = Integer.parseInt(args[3]);
        }

        List<Edge> edges = GraphFileParser.getEdgesFromFile(path_to_graph_data);
        var graph = new Graph(edges);
        List<Integer> path = null;
        if (mode == 0) {
            var djikstra = new DjikstraPathFinder();
            path = djikstra.getShortestPath(graph, source, destination);
        } else if (mode == 1) {
            System.out.println("Nothing to do here");
        } else {
            System.err.println("Incorrect `mode` value. It must be `1` for sequential or `2` for concurrent");
            System.exit(1);
        }
        if (path == null) {
            System.err.println("Can`t build path for such source and destination.");
            System.exit(1);
        }
    }
}

File: implementation/task_4/sources/main.cpp
#include <iostream>
#include "omp.h"
#include "graph.h"
#include "dijkstra.h"


int main(int argc, char *argv[]) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " "
                  << "`path/to/graph.txt`"
                  << "mode: [`1` for sequential dijkstra, "
                  << "`2` for parallel dijkstra]" << std::endl;
        std::exit(1);
    }
    auto graph = Graph(argv[1]);
    Dijkstra *dijkstra;

    int mode = atoi(argv[2]);

    int *distances;
    int source = 0;

    switch (mode) {
        case 1:
            dijkstra = new SequentialDijkstra(&graph);
            distances = dijkstra->calculate_shortest_distances(source);
            break;
        case 2:
            dijkstra = new ParallelDijkstra(&graph);
            distances = dijkstra->calculate_shortest_distances(source);
            break;
        default:
            std::cerr << "Incorrect mode value. It must be [1, 2, 3, 4, 5], but it was " << mode << "." << std::endl;
            std::exit(1);
    }
//    char *filename = "data.txt";
//    dijkstra->save_distances_to_file(filename);
    return 0;
}